code/go/0chain.net/chaincore/chain/state/state_context.go

```diff
@@ -51,19 +51,29 @@ func GetMinter(minter ApprovedMinter) (string, error) {
 *    2) The only from clients valid are txn.ClientID and txn.ToClientID (which will be the smart contract's client id)
  */
 
+type CommonStateContextI interface {
+	GetTrieNode(key datastore.Key, v util.MPTSerializable) error
+	GetBlock() *block.Block
+	GetLatestFinalizedBlock() *block.Block
+}
+
+//go:generate mockery --case underscore --name=QueryStateContextI --output=./mocks
+type QueryStateContextI interface {
+	CommonStateContextI
+	GetEventDB() *event.EventDb
+}
+
 //go:generate mockery --case underscore --name=StateContextI --output=./mocks
 //StateContextI - a state context interface. These interface are available for the smart contract
-// todo this needs to be split up into different interfaces
 type StateContextI interface {
+	QueryStateContextI
```

Comment:
I would suggest not embed the `QueryStateContextI`, as the other `Get*` methods in the `StateContextI` are also `query` like methods, it does not differ from other, but confuse users.

Reply:
See below. I think we need to prevent GetEventDb from being used in smart contracts and InsertTrieNode in queries.